src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs

```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static class SnakeCaseUnitTests
+    {
+        [Fact]
+        public static void ToSnakeCaseTest()
```

Comment:
Based on discussion from the API review, we should add tests for non-ascii text and for text with surrogate pairs (including invalid surrogate pairs) and also validate (at least locally - but I would be ok with checking it in) that the output matches Newtonsoft.Json.

I was working on some testing for CamelCase, so you could try and leverage that as a starting point:
```C#
[Theory]
[InlineData("lowerCase")]
[InlineData("UpperCase")]
[InlineData("UPperCase")]
[InlineData("lower Case")]
[InlineData("loweR Case")]
[InlineData("Upper Case")]
[InlineData("UppeR Case")]
[InlineData("lower case")]
[InlineData("loweR case")]
[InlineData("Upper case")]
[InlineData("UppeR case")]
[InlineData("lower cASe")]
[InlineData("loweR cASe")]
[InlineData("Upper cASe")]
[InlineData("UppeR cASe")]
[InlineData("UPper Case")]
[InlineData("UPpeR Case")]
[InlineData("UPper case")]
[InlineData("UPpeR case")]
[InlineData("UPper cASe")]
[InlineData("UPpeR cASe")]

[InlineData("i")]
[InlineData("I")]
[InlineData("i ")]
[InlineData("I ")]

[InlineData("ii")]
[InlineData("II")]
[InlineData("iI")]
[InlineData("Ii")]
[InlineData("ii ")]
[InlineData("II ")]
[InlineData("iI ")]
[InlineData("Ii ")]

[InlineData("iii")]
[InlineData("III")]
[InlineData("iiI")]
[InlineData("iIi")]
[InlineData("Iii")]
[InlineData("iII")]
[InlineData("IIi")]
[InlineData("IiI")]
[InlineData("iii ")]
[InlineData("III ")]
[InlineData("iiI ")]
[InlineData("iIi ")]
[InlineData("Iii ")]
[InlineData("iII ")]
[InlineData("IIi ")]
[InlineData("IiI ")]

[InlineData("iPhone")]
[InlineData("IPhone")]
[InlineData("IPHone")]
[InlineData("IPH one")]
[InlineData("IPH One")]
[InlineData("IPHONe")]
[InlineData("IPHON e")]
[InlineData("IPHON E")]
[InlineData("IPHONE")]
public static void TestCasing(string baseString)
{
    var testObject = new TestClassWithDictionary
    {
        Dictionary = new Dictionary<string, string>
        {
            { baseString, "hi" }
        }
    };

    string json = JsonSerializer.Serialize<TestClassWithDictionary>(testObject, new JsonSerializerOptions { DictionaryKeyPolicy = JsonNamingPolicy.CamelCase });
    // Assert JSON output is correct

    testObject = JsonSerializer.Deserialize<TestClassWithDictionary>(json);
    json = JsonSerializer.Serialize(testObject);
    // Assert JSON output is correct after round-tripping

    DefaultContractResolver contractResolver = new DefaultContractResolver
    {
        NamingStrategy = new CamelCaseNamingStrategy()
    };

    string expected = JsonConvert.SerializeObject(testObject, new JsonSerializerSettings { ContractResolver = contractResolver });
    // Assert and compare against Newtonsoft.Json

    char[] baseStringArray = baseString.ToCharArray();
    char[] testCharArray = new char[baseStringArray.Length + 2];
    for (int i = 0; i <= baseStringArray.Length; i++)
    {
        for (int j = 0; j < i; j++)
        {
            testCharArray[j] = baseStringArray[j];
        }
        testCharArray[i] = '\uD835';
        testCharArray[i + 1] = '\uDD80';
        //testCharArray[i] = '\uD801';
        //testCharArray[i + 1] = '\uDC27';
        for (int j = i; j < baseStringArray.Length; j++)
        {
            testCharArray[j + 2] = baseStringArray[j];
        }

        string testString = new string(testCharArray);

        testObject = new TestClassWithDictionary
        {
            Dictionary = new Dictionary<string, string>
            {
                { testString, "hi" }
            }
        };

        json = JsonSerializer.Serialize<TestClassWithDictionary>(testObject, new JsonSerializerOptions { DictionaryKeyPolicy = JsonNamingPolicy.CamelCase });
        // Assert JSON output is correct

        testObject = JsonSerializer.Deserialize<TestClassWithDictionary>(json);
        json = JsonSerializer.Serialize(testObject);
        // Assert JSON output is correct after round-tripping

        expected = JsonConvert.SerializeObject(testObject, new JsonSerializerSettings { ContractResolver = contractResolver });
        // Assert and compare against Newtonsoft.Json
    }

    baseStringArray = baseString.ToCharArray();
    testCharArray = new char[baseStringArray.Length + 1];
    for (int i = 0; i <= baseStringArray.Length; i++)
    {
        for (int j = 0; j < i; j++)
        {
            testCharArray[j] = baseStringArray[j];
        }
        testCharArray[i] = '\uD835';
        for (int j = i; j < baseStringArray.Length; j++)
        {
            testCharArray[j + 1] = baseStringArray[j];
        }

        string testString = new string(testCharArray);

        testObject = new TestClassWithDictionary
        {
            Dictionary = new Dictionary<string, string>
            {
                { testString, "hi" }
            }
        };

        json = JsonSerializer.Serialize<TestClassWithDictionary>(testObject, new JsonSerializerOptions { DictionaryKeyPolicy = JsonNamingPolicy.CamelCase });
        // Assert JSON output is correct

        testObject = JsonSerializer.Deserialize<TestClassWithDictionary>(json);
        json = JsonSerializer.Serialize(testObject);
        // Assert JSON output is correct after round-tripping

        expected = JsonConvert.SerializeObject(testObject, new JsonSerializerSettings { ContractResolver = contractResolver });
        // Assert and compare against Newtonsoft.Json
    }
}
```

Reply:
Newtonsoft.Json had some bad translations (`Xml2Json` to `xml2_json`) and the bug I mentioned. Should the behavior be the same or should it be improved a bit? Maybe open an issue in Newtonsoft.Json?

/cc @JamesNK

Reply:
> Should the behavior be the same or should it be improved a bit?

In cases where the Newtonsoft behavior is buggy/undesirable, we should certainly try to make it better here (possibly in both libraries). 

> Maybe open an issue in Newtonsoft.Json?

I'd leave the Newtonsoft.Json side of things to @JamesNK, but filing an issue wouldn't hurt :)

Reply:
Already tracked by JamesNK/Newtonsoft.Json#1956.

Reply:
> Based on discussion from the API review, we should add tests for non-ascii text and for text with surrogate pairs (including invalid surrogate pairs) and also validate (at least locally - but I would be ok with checking it in) that the output matches Newtonsoft.Json.

Can you add some more tests for non-ascii text?